presto-spi/src/main/java/io/prestosql/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -353,77 +352,42 @@ else if (compare < 0) {
      */
     private static long addUnsignedReturnOverflow(Slice left, Slice right, Slice result, boolean resultNegative)
     {
-        // TODO: consider two 7 bytes operations
-        int l0 = getInt(left, 0);
-        int l1 = getInt(left, 1);
-        int l2 = getInt(left, 2);
-        int l3 = getInt(left, 3);
+        long l0 = getLong(left, 0);
+        long l1 = getLong(left, 1);
 
-        int r0 = getInt(right, 0);
-        int r1 = getInt(right, 1);
-        int r2 = getInt(right, 2);
-        int r3 = getInt(right, 3);
+        long r0 = getLong(right, 0);
+        long r1 = getLong(right, 1);
 
-        long intermediateResult;
-        intermediateResult = toUnsignedLong(l0) + toUnsignedLong(r0);
+        long z0 = l0 + r0;
+        // Long.unsignedCompare() inlined
+        // (unsigned) z0 < (unsigned) l0
+        int overflow = z0 + Long.MIN_VALUE < l0 + Long.MIN_VALUE ? 1 : 0;
 
-        int z0 = (int) intermediateResult;
+        long intermediateResult = l1 + r1 + overflow;
+        long z1 = intermediateResult & (~SIGN_LONG_MASK);
+        pack(result, z0, z1, resultNegative);
 
-        intermediateResult = toUnsignedLong(l1) + toUnsignedLong(r1) + (intermediateResult >>> 32);
-
-        int z1 = (int) intermediateResult;
-
-        intermediateResult = toUnsignedLong(l2) + toUnsignedLong(r2) + (intermediateResult >>> 32);
-
-        int z2 = (int) intermediateResult;
-
-        intermediateResult = toUnsignedLong(l3) + toUnsignedLong(r3) + (intermediateResult >>> 32);
-
-        int z3 = (int) intermediateResult & (~SIGN_INT_MASK);
-
-        pack(result, z0, z1, z2, z3, resultNegative);
-
-        return intermediateResult >> 31;
+        return intermediateResult >>> 63;
     }
 
     /**
      * This method ignores signs of the left and right and assumes that left is greater then right
      */
     private static void subtractUnsigned(Slice left, Slice right, Slice result, boolean resultNegative)
     {
-        // TODO: consider two 7 bytes operations
-        int l0 = getInt(left, 0);
-        int l1 = getInt(left, 1);
-        int l2 = getInt(left, 2);
-        int l3 = getInt(left, 3);
-
-        int r0 = getInt(right, 0);
-        int r1 = getInt(right, 1);
-        int r2 = getInt(right, 2);
-        int r3 = getInt(right, 3);
+        long l0 = getLong(left, 0);
+        long l1 = getLong(left, 1);
 
-        long intermediateResult;
-        intermediateResult = toUnsignedLong(l0) - toUnsignedLong(r0);
+        long r0 = getLong(right, 0);
+        long r1 = getLong(right, 1);
 
-        int z0 = (int) intermediateResult;
+        long z0 = l0 - r0;
+        // Long.unsignedCompare() inlined
```

Comment:
similar here, it's not `unsignedCompare` inlined. Please extract a separate method and add a comment what it's based on.